### DISCUSSION TITLE
How do I do the first initial (module) navigation based on some condition?

### DISCUSSION THREAD

**Discussion by nivle** [PrismLibrary/Prism#2844] [WPF] [answered]
Hi

I have 3 modules, that use the same region.
And the first module only needs to be shown on a condition, else the second module is shown.

So I added my modules to the catalog like this, in my app.

```
        protected override Window CreateShell()
        {
            return Container.Resolve<MainWindow>();
        }

        protected override async void RegisterTypes(IContainerRegistry containerRegistry)
        {

        }

        protected override void ConfigureModuleCatalog(IModuleCatalog moduleCatalog)
        {
            moduleCatalog.AddModule<InitialSettingsModule>();
            moduleCatalog.AddModule<HomeModule>();
            moduleCatalog.AddModule<SettingsModule>();
        }
```

And then I did try to navigate inside my viewmodel contructor of initialsettings, but then nothing happens as described here: https://github.com/PrismLibrary/Prism/discussions/2586

So now I'm confused, where am I supposed to place the conditional code to navigate to homemodule?

I've also tried adding the condition in my app like this:

```
        private IContainerRegistry containerRegistry;

        protected override Window CreateShell()
        {
            return Container.Resolve<MainWindow>();
        }

        protected override async void RegisterTypes(IContainerRegistry containerRegistry)
        {
            this.containerRegistry = containerRegistry;
        }

        protected override async void ConfigureModuleCatalog(IModuleCatalog moduleCatalog)
        {
            IInitialSettingsService initialSettingsService = containerRegistry.GetContainer().GetRequiredService<IInitialSettingsService>();

            if (!await initialSettingsService.AreInitialSettingsSetAsync())
            {
                moduleCatalog.AddModule<InitialSettingsModule>();
            }

            moduleCatalog.AddModule<HomeModule>();
            moduleCatalog.AddModule<SettingsModule>();
        }
```

But this resulted in a black screen when I ran the app, which probably has to do something with the fact that this line of code is async, as it's no longer a black screen when I comment it out.

`
        if (!await initialSettingsService.AreInitialSettingsSetAsync())
`

So now I'm not sure where to add the condition, or if this is even possible?

**Comment by brianlagunas** [comment, marked as answer]
Anything is possible. It's just code. The reason navigation isn't working in your InitialSettingsModule is because you are trying to navigate to a view in a module that hasn't been loaded yet. So, the view doesn't exist yet. You can only navigate to views in other modules if those modules have been loaded. You just have to make sure all modules that contain the views you want to navigate to are loaded first. However, it sounds like you're trying to conditionally load your modules, not views. Why not wrap that logic around the module itself and only add it if meets your condition?

**Reply by nivle** [reply]
Thanks for your reply, I ended up moving the condition to the MainWindowViewModel, that's where all the modules, and views seem to be loaded, so the condition does work there, so I can do my conditional initial navigation there.

**Comment by nivle** [comment]
I solved it for now by adding the condition in the MainWindowViewModel and navigating there

so my app now looks like this


```
        protected override Window CreateShell()
        {
            return Container.Resolve<MainWindow>();
        }

        protected override async void RegisterTypes(IContainerRegistry containerRegistry)
        {

        }

        protected override void ConfigureModuleCatalog(IModuleCatalog moduleCatalog)
        {
            moduleCatalog.AddModule<InitialSettingsModule>();
            moduleCatalog.AddModule<HomeModule>();
            moduleCatalog.AddModule<SettingsModule>();
        }
```

and my MainWindowViewModel looks like this (**I'm using CommunityToolkit.Mvvm for my commands**)


```
        private readonly IRegionManager regionManager;
        private readonly IInitialSettingsService initialSettingsService;

        public string Title { get; set; } = "test";

        public MainWindowViewModel(IRegionManager regionManager, IInitialSettingsService initialSettingsService)
        {
            this.regionManager = regionManager;
            this.initialSettingsService = initialSettingsService;

            InitialNavigateCommand.Execute(new NavigationRoute() { Region = RegionNames.ContentRegion, Route = "Home" });
        }

        [RelayCommand]
        private async Task InitialNavigate(NavigationRoute navigationRoute)
        {
            if (await initialSettingsService.AreInitialSettingsSetAsync())
            {
                Navigate(navigationRoute);
            }
        }

        [RelayCommand]
        private void Navigate(NavigationRoute navigationRoute)
        {
            regionManager.RequestNavigate(navigationRoute.Region, navigationRoute.Route);
        }
```
